Issue458 add carousel example

jongund opened the pull request.

jongund commented:
@mcking65 I have updated the regression tests so they pass for carousel-1 example.  I have added a few more regression tests, but there are still more to add.

jongund commented:
@mcking65 I updated the example to include roledescription and added regression test for roledescription

mcking65 commented on examples/carousel/carousel-1/js/carousel.js:
```diff
@@ -0,0 +1,209 @@
+    this.liveRegionNode.setAttribute('aria-live', 'off');
+    this.pauseButton.innerHTML = this.stopLabel;
+  }
+  this.pauseButton.disabled = this.hasHover | this.hasFocus;
```
@jongund, I think we should use aria-disabled and keep the button focusable when disabled so SR users can hear the state of the carousel when tabbing.

mcking65 commented on examples/carousel/carousel-1/js/carousel.js:
```diff
@@ -0,0 +1,209 @@
+  this.pauseButton.disabled = this.hasHover | this.hasFocus;
+  this.rotate = false;
+  this.liveRegionNode.setAttribute('aria-live', 'polite');
+  this.pauseButton.innerHTML = this.startLabel;
```
Think you want this to be playLabel, not startLabel. Button label is undefined when carousel is stopped.

mcking65 commented on examples/carousel/carousel-1/js/carousel.js:
```diff
@@ -0,0 +1,209 @@
+  this.currentItem = null;
+  this.pauseButton = null;
+
+  this.playLabel = 'Play Slides';
```
I'm still advocating for a more descriptive label that I think will be more broadly understood. I don't believe most people would understand what this label means.

mcking65 commented on examples/carousel/carousel-1/carousel-1.html:
```diff
@@ -0,0 +1,577 @@
+
+                    <div class="carousel-caption">
+
+                      <h2>
```
@jongund, please change the h2 elements inside the slides to h3 so they fit in the page IA.

jongund commented:
@mcking65 I have pushed the following changes:
* Changed h2s to h3s in the slides
* Changed using `button.disabled` property to `aria-disabled`
* Changed pause button labels to "Start automatic slide rotation" and "Stop automatic slide rotation"
* Added disabled property to documentation

mcking65 commented:
@jongund commented:
> @mcking65 I have pushed the following changes:
> 
> * Changed h2s to h3s in the slides

Looks good.

> * Changed using `button.disabled` property to `aria-disabled`

Ditto

> * Changed pause button labels to "Start automatic slide rotation" and "Stop automatic slide rotation"

Good, although still waiting for input from others on the mailing list. This is easy to adjust at any time if the group decides on something different.

> * Added disabled property to documentation

Thank you. I'll now finish up my part of the documentation work.

charmarkk commented on examples/carousel/carousel-1/carousel-1.html:
```diff
@@ -0,0 +1,533 @@
+              </li>
+              <li>
+                Moving keyboard focus to any of the carousel content, including the next and previous slide elements, pauses automatic rotation.
+                Automatic rotation resumes when keyboard focus moves out of the carousel content unles another condition, such as mouse hover,  that prevents rotation has been triggered.
```
"unles" to "unless"

charmarkk commented on examples/carousel/carousel-1/carousel-1.html:
```diff
@@ -0,0 +1,533 @@
+                  <li>The button is hidden off-screen when it does not have keyboard focus, and becomes visible when users move focus to it with <kbd>Tab</kbd>.</li>
+                  <li>
+                    When not visible, the button is positioned off-screen, instead of being hidden with <code>display:none</code>, so it is always available to screen reader users.
+                    This is an important aspect of supporting users of screen readers that do not move the mouse or keyboard focus as the user is reading.
```
What is this line trying to say? It seems a little clunky. Are we saying something like "This is an important aspect of supporting users of screen readers that **the carousel does** not move the mouse or keyboard focus as the user is reading." ?

charmarkk replied:
Ah now I see, that makes sense. I thought that may have been it, but it wasn't preciseely clear. Maybe we change the second "the user" to "they" since we already preface with "...users that..."? 

If other folks think it's clear, that's fine! I understand now.

mcking65 replied:
> Ah now I see, that makes sense. I thought that may have been it, but it wasn't preciseely clear. Maybe we change the second "the user" to "they" since we already preface with "...users that..."?
> 
> If other folks think it's clear, that's fine! I understand now.


In commit 3d920fc  I changed that sentence to:

>This ensures the rotation control is accessible to people using a screen reader that does not move the mouse or keyboard focus when the user is moving its reading cursor.

mcking65 replied:
@annabbott, maybe you didn't see the html file on the files tab because it had lots of lines changed. When the diff is large, github does not load it by default. You have to click a button to show the changed lines in the file.

charmarkk replied:
@mcking65's commit 3d920fc solves my issue - my thanks, Matt!

annabbott commented:
I can't figure out how to make edit suggestions to the text - when I go to Files, all I see is code.

Second sentence of second paragraph following h1
 For instance, Rotation stops when users move focus...
Lower case "r" on "rotation".

The rest looks good to me.
NOTE: I did NOT test with an AT.

jnurthen commented on examples/carousel/carousel-1/carousel-1.html:
```diff
@@ -0,0 +1,533 @@
+        The following example implementation of the
+        <a href="../../../#carousel">carousel design pattern</a>
+        demonstrates features of the pattern that are essential to accessibility for carousels that automatically start rotating when the page loads.
+        For instance, Rotation stops when users move focus to any control or link in the carousel or hover the mouse over carousel content.
```
```suggestion
        For instance, rotation stops when users move focus to any control or link in the carousel or hover the mouse over carousel content.
```

jnurthen commented:
> Second sentence of second paragraph following h1
> For instance, Rotation stops when users move focus...
> Lower case "r" on "rotation".
@annabbott I added a review comment for you

mcking65 commented on examples/carousel/carousel-1/carousel-1.html:
```diff
@@ -239,7 +239,7 @@ <h2>Accessibility Features</h2>
                     This is an important aspect of supporting users of screen readers that do not move the mouse or keyboard focus as the user is reading.
                   </li>
-                  <li>If the carousel is rotating, the button is labeled "Stop Automatic Slide Show", informing screen reader users that the slides are changing in addition to providing a way to stop the changes.</li>
+                  <li>If the carousel is rotating, the button is labeled "Stop Automatic Slide Show," informing screen reader users that the slides are changing in addition to providing a way to stop the changes.</li>
```
@charmarkk, we may want to research this convention so we have a consistent approach. For most of eternity, the Chicago Manual of Style, which is referenced by W3C editorial guidelines, has specified that punctuation belongs inside closing quotes.

However, within recent decades, I thought some exceptions were introduced into the Chicago Manual of Style. In this case, the quotes are denoting a precise string that is used as the label. The actual label does not include the punctuation, and we would not want it to include the punctuation. This is the type of circumstance where I thought the manual provided flexibility for the position of the punctuation.

We have it both ways in the APG. It would be good to have a consistent approach going forward. It would be helpful if someone could find the relevant rules.

charmarkk commented:
I can take on researching that for us!

Mark McCarthy
________________
Sent from my mobile ?

Please excuse minor errors, I use dictation when mobile.

________________________________
From: Matt King <notifications@github.com>
Sent: Friday, January 11, 2019 8:17:15 PM
To: w3c/aria-practices
Cc: Mark McCarthy; Mention
Subject: Re: [w3c/aria-practices] Add carousel example (#958)


@mcking65 commented on this pull request.

________________________________

In examples/carousel/carousel-1/carousel-1.html<https://github.com/w3c/aria-practices/pull/958#discussion_r247299059>:

> @@ -239,7 +239,7 @@ <h2>Accessibility Features</h2>
                     When not visible, the button is positioned off-screen, instead of being hidden with <code>display:none</code>, so it is always available to screen reader users.
                     This is an important aspect of supporting users of screen readers that do not move the mouse or keyboard focus as the user is reading.
                   </li>
-                  <li>If the carousel is rotating, the button is labeled "Stop Automatic Slide Show", informing screen reader users that the slides are changing in addition to providing a way to stop the changes.</li>
+                  <li>If the carousel is rotating, the button is labeled "Stop Automatic Slide Show," informing screen reader users that the slides are changing in addition to providing a way to stop the changes.</li>


@charmarkk<https://github.com/charmarkk>, we may want to research this convention so we have a consistent approach. For most of eternity, the Chicago Manual of Style, which is referenced by W3C editorial guidelines, has specified that punctuation belongs inside closing quotes.

However, within recent decades, I thought some exceptions were introduced into the Chicago Manual of Style. In this case, the quotes are denoting a precise string that is used as the label. The actual label does not include the punctuation, and we would not want it to include the punctuation. This is the type of circumstance where I thought the manual provided flexibility for the position of the punctuation.

We have it both ways in the APG. It would be good to have a consistent approach going forward. It would be helpful if someone could find the relevant rules.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/w3c/aria-practices/pull/958#pullrequestreview-191922711>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AVO5zUT2C7vAKZbafwysAU3edjya2MWmks5vCUWrgaJpZM4ZcVTk>.

jongund commented:
@mcking65 Removed the use of aria-describedby from each slide and updated the regression test to use new ids

accdc commented:
Hi,
It looks good to me. As we spoke of during the call, minus the aria-describedby attribute.

As discovered during testing, screen reader support for aria-roledescription is still incomplete. As such, the screen reader support of this design pattern is not sufficient to justify mainstream developers to use this as part of public web technologies, but for now, this widget is something that screen reader providers can use to test against to make sure this functionality is built in as time goes on.

mcking65 commented:
All review comments have been addressed. Completion of regression test development is tracked by issue #969. Merging as discussed in today's meeting. Thank you all, especially @jongund, for helping bring this to life!

shirsha commented:
The content over the image, and previous/next button are difficult to see due to the busyness of the image behind it. Add a background color to the buttons and insert CSS color block between the text and the image so that they can be visible even when the background image is busy.

elamurugan-r commented:
a)
```
<section aria-roledescription="carousel" aria-labelledby="carousel-header" >
        <h1 id="carousel-header" >Popular Web series and Episodes</h1>
        <button>Trigger</button>
</section>
```
b)
```
<div role="region" aria-roledescription="carousel" aria-labelledby="carousel-header" >
        <h1 id="carousel-header" >Popular Web series and Episodes</h1>
        <button>Trigger</button>
</div>
```
 For the above codes, axe-core-4.3.5 gives the following error eventhough using implicit/explicit aria-role: 
a & b) **aria-roledescription must be on elements with a semantic role.**
 **_Issue Description:_ Ensure aria-roledescription is only used on elements with an implicit or explicit role**

Screen reader (ChromeVox) is reading the above as
**a) Section Trigger button**
**b) Popular Web Series and its Episodes Trigger**

Its not reading the text inside aria-roledescription.
